sys/pm_callouts.c

```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
+                    packetInfo->interfaceIndex, packetInfo->subInterfaceIndex,
+                    injectNBL, free_after_inject,
+                    tcpResetPacket);
+        }
+
+        if (!NT_SUCCESS(status)) {
+            ERR("Failed to inject IPv6/TCP RST packet: 0x%x", status);
+            free_after_inject(tcpResetPacket, injectNBL, FALSE);
+        }  else {
+            INFO("IPv6/TCP RST packet was send as a response to: %s", print_packet_info(packetInfo));
+        }
+    } else {
+        // Initialize header for the original packet with SYN flag
+        ULONG originalIPHeaderLength = calc_ipv4_header_size(originalPacket, originalPacketLength);
+        PIPV4_HEADER originalIPHeader = (PIPV4_HEADER) originalPacket;
+        PTCP_HEADER originalTCPHeader = (PTCP_HEADER) ((UINT8*)originalPacket + originalIPHeaderLength);
+        
+        // Initialize variables
+        PNET_BUFFER_LIST injectNBL;
+        NTSTATUS status;
+        UINT16 packetLength = sizeof(IPV4_HEADER) + sizeof(TCP_HEADER);
+        void *tcpResetPacket;
+        PIPV4_HEADER ipHeader;
+        PTCP_HEADER tcpHeader;
+
+        // allocate memory for the reset packet
+        tcpResetPacket = portmaster_malloc(packetLength, FALSE);
+
+        // initialize IPv4 header
+        ipHeader = (PIPV4_HEADER) tcpResetPacket;        
+        ipHeader->HdrLength = sizeof(IPV4_HEADER) / 4;
+        ipHeader->Version = 4;
+        ipHeader->TOS = 0;
+        ipHeader->Length = RtlUshortByteSwap(packetLength);
+        ipHeader->Id = 0;
+        ipHeader->Protocol = packetInfo->protocol;
+        ipHeader->TTL = 128;
+        ipHeader->DstAddr = originalIPHeader->SrcAddr; // Source becomes destination
+        ipHeader->SrcAddr = originalIPHeader->DstAddr; // Destination becomes source
+
+         // Initialize TCP header
+        tcpHeader = (PTCP_HEADER) ((UINT8*)tcpResetPacket + sizeof(IPV4_HEADER));
+        tcpHeader->SrcPort = originalTCPHeader->DstPort; // Source becomes destination
+        tcpHeader->DstPort = originalTCPHeader->SrcPort; // Destination becomes source
+        tcpHeader->HdrLength = sizeof(TCP_HEADER) / 4;
+        tcpHeader->SeqNum = 0;
+        // We should acknowledge the SYN packet while doing the reset
+        tcpHeader->AckNum = RtlUlongByteSwap(RtlUlongByteSwap(originalTCPHeader->SeqNum) + 1);
+        tcpHeader->Ack = 1;
+        tcpHeader->Rst = 1;
+
+        calc_ipv4_checksum(tcpResetPacket, packetLength, TRUE);
+
+        // Create the net buffer list for the new RST packet
+        status = wrap_packet_data_in_nb(tcpResetPacket, packetLength, &injectNBL);
+        if (!NT_SUCCESS(status)) {
+            ERR("send_tcp_rst ipv4 -> wrap_packet_data_in_nb failed: %u", status);
+            portmaster_free(tcpResetPacket);
+            return;
+        }
+
+        // loopback needs special treatment
+        if(is_ipv4_loopback(packetInfo->remoteIP[0])) {
```

Comment:
This _should_ only be happening for outgoing packets, but we may want to block an incoming packet.
I suggest to check for loopback, or incoming packets.